samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java

```diff
@@ -290,6 +242,175 @@ public void updatesAndReads(
         NUM_PROCESSORS / 2);
   }
 
+  public void doWorkSingleTransactionParallelLocking(DatabaseClient dbClient) {
+    // getting the available processor
+    ExecutorService executorService = Executors.newFixedThreadPool(NUM_PROCESSORS);
+
+    List<Future<Integer>> futures = new ArrayList<>();
+    // List to store the count of total number of workItem(row) which has not yet completed means
+    // is_done is false.
+    List<String> workItemCount = new ArrayList<>();
+    // Below logic is reading the generated_value for the rows which are having is_done=false. We
+    // have added generated_value column value in incremented order starting from 1 in the table.
+    // This will help us to assign executing the random workItem(rows) work without overlapping. As
+    // each thread will pick up different generated_value to query.
+    try (ResultSet resultSet =
+        dbClient
+            .singleUse() // Execute a single read or query against Cloud Spanner.
+            // Below query can be parameterized to but kept here for simplicity
+            .executeQuery(
+                Statement.of("SELECT generated_value FROM WorkList WHERE is_done is false"))) {
+      while (resultSet.next()) {
+        workItemCount.add(resultSet.getString(0));
+      }
+    }
+
+    if (workItemCount.size() == 0) {
+      System.out.println(
+          "There is no workItem left to process, please update the is_done value to false, or add more work item in the table");
+      return;
+    }
+
+    // Assigning each thread the amount of work equally
+    for (int i = 0; i < NUM_PROCESSORS; ++i) {
+      // Each thread get the index from this loop, total workItem to be processed and number of
+      // thread (numProcessors) available for processing.
+      // E.g In case of 10 threads and 30 workItem the assignment for this method would be like
+
+      // Thread 1: ThreadIndex(i) 0, workitem 30
+      // Thread 2: ThreadIndex(i) 1  workitem 30
+      // Thread 3: ThreadIndex(i) 2  workitem 30
+      // ...
+      // Thread 10: ThreadIndex(i) 9  workitem 30
+      // These values will be determined by received by each thread(an object of WorkItemProcessor)
+      // while executing based on the input values passed here.
+      String lockQueryInTransaction =
+          "SELECT * FROM WorkList WHERE is_done is false and generated_value= '%s';";
```

Comment:
The SQL command that you should run is "SELECT * FROM WorkList WHERE is_done is false LIMIT 100", and then use a random number to pick from the rows returned which WorkItem to do. This should let you remove generated_value, as well as remove the precompute step where you store all the workitems locally.

Reply:
What I understand is Query-> "SELECT * FROM WorkList WHERE is_done is false LIMIT 100" will run outside of any transaction(the way I ran above one but in on method) , then pick a row randomly out of those 100 and use a query to fetch data for that specific row in ReadwriteTransaction(). If **is_done** is still false update it in same transaction else exit and pick up any thing new.

Reply:
This is true for the second "correct" method. For the first method, it should run as part of the read/write transaction.